File: src/data/sprites.ts
/** ROUNDrBR.VH0 vehicle sprites: 32-direction rotation table as built by fn 4611/4696/46f6.
 *  Port of tools/mm/sprites.py; layout in re/notes/61-sprites.md. */
export interface VehicleSprites {
  size: number;            // 24 or 40
  frames: Uint8Array[];    // 32 frames (size*size each), index = heading >> 3
  extra: Uint8Array[];     // extra sprites copied to segment 5D78 (12 x 24x24 or 5 x 40x40)
  rawFrames: number;       // frames present in the decoded file
}

function vflip(f: Uint8Array, w: number): Uint8Array {
  const o = new Uint8Array(w * w);
  for (let r = 0; r < w; r++) o.set(f.subarray((w - 1 - r) * w, (w - r) * w), r * w);
  return o;
}
function hflip(f: Uint8Array, w: number): Uint8Array {
  const o = new Uint8Array(w * w);
  for (let r = 0; r < w; r++) for (let x = 0; x < w; x++) o[r * w + x] = f[r * w + (w - 1 - x)]!;
  return o;
}

/** 9 source frames (0..90 degrees) -> 32 frames exactly like fn 4696 (24) / fn 46f6 (40). */
export function buildFrames(base9: Uint8Array[], w: number): Uint8Array[] {
  const fr: Uint8Array[] = [...base9];
  for (let i = 7; i >= 0; i--) fr.push(vflip(base9[i]!, w));          // 9..16  = vflip(7..0)
  for (let i = 15; i >= 9; i--) fr.push(hflip(fr[i]!, w));            // 17..23 = hflip(15..9)
  for (let i = 8; i >= 1; i--) fr.push(hflip(base9[i]!, w));          // 24..31 = hflip(8..1)
  return fr;
}

export function decodeVehicle(data: Uint8Array, round: number): VehicleSprites {
  const w = round === 9 ? 40 : 24;
  const fsz = w * w;
  const nExtra = round === 9 ? 5 : 12;
  const need = (9 + nExtra) * fsz;
  const buf = new Uint8Array(Math.max(need, data.length)); buf.set(data);   // past the data: stale window, zero here
  const base9: Uint8Array[] = [];
  for (let i = 0; i < 9; i++) base9.push(buf.slice(i * fsz, (i + 1) * fsz));
  const extra: Uint8Array[] = [];
  for (let i = 0; i < nExtra; i++) extra.push(buf.slice((9 + i) * fsz, (10 + i) * fsz));
  return { size: w, frames: buildFrames(base9, w), extra, rawFrames: Math.floor(data.length / fsz) };
}

/** Exact image of segment 4D78 after loading (0x4800 B for 24x24, 0xC800 B for 40x40). */
export function frameTableBytes(v: VehicleSprites): Uint8Array {
  const out = new Uint8Array(v.frames.length * v.size * v.size);
  v.frames.forEach((f, i) => out.set(f, i * v.size * v.size));
  return out;
}

File: tools/mm/sprites.py
"""ROUNDrBR.VH0 vehicle sprites: rotation table as built by MICRO.EXE fn 4611/4696/46f6 (see re/notes/61-sprites.md)."""
from dataclasses import dataclass

@dataclass
class VehicleSprites:
    size: int            # 24 or 40
    frames: list         # 32 rotation frames (bytes, size*size each), index = heading >> 3
    extra: list          # extra sprites copied to segment 5D78 (12 x 24x24 or 5 x 40x40)
    raw_frames: int      # frames present in the decoded file

def _vflip(f, w):
    return b''.join(f[r * w:(r + 1) * w] for r in range(w - 1, -1, -1))

def _hflip(f, w):
    return b''.join(f[r * w:(r + 1) * w][::-1] for r in range(w))

def build_frames(base9, w):
    """9 source frames (0..90 deg) -> 32 frames exactly like fn 4696 (24) / fn 46f6 (40)."""
    fr = list(base9)
    fr += [_vflip(base9[i], w) for i in range(7, -1, -1)]          # 9..16  = vflip(7..0)
    fr += [_hflip(fr[i], w) for i in range(15, 8, -1)]              # 17..23 = hflip(15..9)
    fr += [_hflip(base9[i], w) for i in range(8, 0, -1)]            # 24..31 = hflip(8..1)
    assert len(fr) == 32
    return fr

def decode_vehicle(data: bytes, round_no: int) -> VehicleSprites:
    w = 40 if round_no == 9 else 24
    fsz = w * w
    n_rot = 9
    n_extra = 5 if round_no == 9 else 12
    need = (n_rot + n_extra) * fsz
    buf = data + bytes(max(0, need - len(data)))    # the game copies past the decoded data (stale LZ window); zero here
    base9 = [buf[i * fsz:(i + 1) * fsz] for i in range(n_rot)]
    extra = [buf[(n_rot + i) * fsz:(n_rot + i + 1) * fsz] for i in range(n_extra)]
    return VehicleSprites(w, build_frames(base9, w), extra, len(data) // fsz)

def frame_table_bytes(v: VehicleSprites) -> bytes:
    """Exact image of segment 4D78 after loading (0x4800 B for 24x24, 0xC800 B for 40x40)."""
    return b''.join(v.frames)
